builtin/logical/pki/crl_util.go

```diff
@@ -1550,6 +1767,87 @@ func getRevokedCertEntries(sc *storageContext, issuerIDCertMap map[issuerID]*x50
+	// Getting unified revocation entries is a bit different than getting
+	// the local ones. In particular, the full copy of the certificate is
+	// unavailable, so we'll be able to avoid parsing the stored certificate,
+	// at the expense of potentially having incorrect issuer mappings.
+	var unassignedCerts []pkix.RevokedCertificate
+	revokedCertsMap := make(map[issuerID][]pkix.RevokedCertificate)
+
+	listingPath := unifiedRevocationReadPathPrefix
+	if isDelta {
+		listingPath = unifiedDeltaWALPrefix
+	}
+
+	// First, we find all clusters that have written certificates.
+	clusterIds, err := sc.Storage.List(sc.Context, listingPath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to list clusters for unified CRL building: %w", err)
+	}
+
+	// Then for every cluster, we find its revoked certificates...
+	for _, clusterId := range clusterIds {
+		if !strings.HasSuffix(clusterId, "/") {
+			// No entries
+			continue
+		}
+
+		clusterPath := listingPath + clusterId
+		serials, err := sc.Storage.List(sc.Context, clusterPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to list serials in cluster (%v) for unified CRL building: %w", clusterId, err)
+		}
+
+		// At this point, we need the storage entry. Rather than using the
+		// clusterPath and adding the serial, we need to use the true
+		// cross-cluster revocation entry (as, our above listing might have
+		// used delta WAL entires without the full revocation info).
+		serialPrefix := unifiedRevocationReadPathPrefix + clusterId
+		for _, serial := range serials {
+			if isDelta && (serial == deltaWALLastBuildSerialName || serial == deltaWALLastRevokedSerialName) {
+				// Skip our placeholder entries...
+				continue
+			}
+
+			serialPath := serialPrefix + serial
+			entryRaw, err := sc.Storage.Get(sc.Context, serialPath)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to read unified revocation entry in cluster (%v) for unified CRL building: %w", clusterId, err)
+			}
+			if entryRaw == nil {
+				// Skip empty entries. We'll eventually tidy them.
+				continue
+			}
+
+			var xRevEntry unifiedRevocationEntry
+			if err := entryRaw.DecodeJSON(&xRevEntry); err != nil {
+				return nil, nil, fmt.Errorf("failed json decoding of unified revocation entry at path %v: %w ", serialPath, err)
+			}
+
+			// Convert to pkix.RevokedCertificate entries.
+			var revEntry pkix.RevokedCertificate
+			var ok bool
+			revEntry.SerialNumber, ok = serialToBigInt(serial)
+			if !ok {
+				return nil, nil, fmt.Errorf("failed to encode serial for CRL building: %v", serial)
+			}
+
+			revEntry.RevocationTime = xRevEntry.RevocationTimeUTC
+
+			// Finally, add it to the correct mapping.
+			_, present := issuerIDCertMap[xRevEntry.CertificateIssuer]
+			if present {
+				unassignedCerts = append(unassignedCerts, revEntry)
+			} else {
+				revokedCertsMap[xRevEntry.CertificateIssuer] = append(revokedCertsMap[xRevEntry.CertificateIssuer], revEntry)
```

Comment:
If we have duplicate serials across clusters, I *believe* we will generate a CRL with duplicates as well?

Reply:
Ah, this is fair, I should address that. I think perhaps we maintain a map of seen serials separately?